ukvm/ukvm_hv_openbsd_x86_64.c

```diff
@@ -0,0 +1,254 @@
+        ukvm_gpa_t gpa_kend, char **cmdline)
+{
+    struct ukvm_hvb *hvb = hv->b;
+
+    struct vm_resetcpu_params vrp = {
+        .vrp_vm_id = hvb->vcp_id,
+        .vrp_vcpu_id = hvb->vcpu_id,
+        .vrp_init_state = {
+            .vrs_gprs[VCPU_REGS_RFLAGS] = X86_RFLAGS_INIT,
+            .vrs_gprs[VCPU_REGS_RIP] = gpa_ep,
+            .vrs_gprs[VCPU_REGS_RSP] = hv->mem_size - 8,
+            .vrs_gprs[VCPU_REGS_RDI] = X86_BOOT_INFO_BASE,
+            .vrs_crs[VCPU_REGS_CR0] = X86_CR0_INIT,
+            .vrs_crs[VCPU_REGS_CR3] = X86_CR3_INIT,
+            .vrs_crs[VCPU_REGS_CR4] = (X86_CR4_INIT | X86_CR4_VMXE),
+            .vrs_sregs[VCPU_REGS_CS] = sreg_to_vsi(&ukvm_x86_sreg_code),
+            .vrs_sregs[VCPU_REGS_DS] = sreg_to_vsi(&ukvm_x86_sreg_data),
+            .vrs_sregs[VCPU_REGS_ES] = sreg_to_vsi(&ukvm_x86_sreg_data),
+            .vrs_sregs[VCPU_REGS_FS] = sreg_to_vsi(&ukvm_x86_sreg_data),
+            .vrs_sregs[VCPU_REGS_GS] = sreg_to_vsi(&ukvm_x86_sreg_data),
+            .vrs_sregs[VCPU_REGS_SS] = sreg_to_vsi(&ukvm_x86_sreg_data),
+            .vrs_gdtr = { 0x0, X86_GDTR_LIMIT, 0x0, X86_GDT_BASE},
+            .vrs_idtr = { 0x0, 0xFFFF, 0x0, 0x0},
+            .vrs_sregs[VCPU_REGS_LDTR] = sreg_to_vsi(&ukvm_x86_sreg_unusable),
+            .vrs_sregs[VCPU_REGS_TR] = sreg_to_vsi(&ukvm_x86_sreg_tr),
+            .vrs_msrs[VCPU_REGS_EFER] = X86_EFER_INIT,
+            .vrs_msrs[VCPU_REGS_STAR] = 0ULL,
+            .vrs_msrs[VCPU_REGS_LSTAR] = 0ULL,
+            .vrs_msrs[VCPU_REGS_CSTAR] = 0ULL,
+            .vrs_msrs[VCPU_REGS_SFMASK] = 0ULL,
+            .vrs_msrs[VCPU_REGS_KGSBASE] = 0ULL,
+            .vrs_crs[VCPU_REGS_XCR0] = XCR0_X87
+        }
+    };
+
+    ukvm_x86_setup_gdt(hv->mem);
+    ukvm_x86_setup_pagetables(hv->mem, hv->mem_size);
+
+    struct ukvm_boot_info *bi =
+        (struct ukvm_boot_info *)(hv->mem + X86_BOOT_INFO_BASE);
+    bi->mem_size = hv->mem_size;
+    bi->kernel_end = gpa_kend;
+    bi->cmdline = X86_CMDLINE_BASE; 
+    bi->cpu.tsc_freq = get_tsc_freq();
+
+    if (ioctl(hvb->vmd_fd, VMM_IOC_RESETCPU, &vrp) < 0)
+        err(1, "Cannot reset VCPU - exiting.");
+
+    *cmdline = (char *)(hv->mem + X86_CMDLINE_BASE);
+}
+
+void ukvm_hv_vcpu_loop(struct ukvm_hv *hv)
+{
+    struct ukvm_hvb         *hvb = hv->b;
+    struct vm_run_params    *vrp;
+    struct passwd *pw;
+    uid_t uid;
+    gid_t gid;
+
+    vrp = malloc(sizeof(struct vm_run_params));
+    if (vrp == NULL)
+        err(1, "calloc vrp");
+
+    vrp->vrp_exit = malloc(sizeof(union vm_exit));
+    if (vrp == NULL)
+        err(1, "calloc vrp_exit");
+
+    vrp->vrp_vm_id = hvb->vcp_id;
+    vrp->vrp_vcpu_id = hvb->vcpu_id;
+    vrp->vrp_continue = 0;
+
+    if ((pw = getpwnam(VMD_USER)) == NULL)
+        err(1, "can't get _vmd user");
+    uid = pw->pw_uid;
+    gid = pw->pw_gid;
+
+    if (chroot(pw->pw_dir) == -1)
+        err(1, "chroot: %s", pw->pw_dir);
+
+    if (chdir("/") == -1)
```

Comment:
Don't think this is needed (at least it's not in Linux).

Reply:
Its not needed, but it is much more secure chroot'ing ukvm.

Reply:
sounds good

Reply:
With the Linux/FreeBSD ports we've been leaving any extra security (chroot, jails, containers, whatever) up to the user for now. If its the OpenBSD way to always do this in every tool then this is fine. (We can always change/refactor it later if we end up doing more of this ourselves also on other systems)

Reply:
It is the OpenBSD way to do this with most, if not all tools.